# Align to chimera
--------------------------------------------------

## Table of Contents
- [Align to chimera](#align-to-chimera)
  - [Table of Contents](#table-of-contents)
  - [Project overview](#project-overview)
  - [Content](#content)
  - [Quick start](#quick-start)

--------------------------------------------------

## Project overview 

This repository contains all the necessary scripts to perform a calibrated whole-genome alignment of paired-end genomic bulk-sequencing data. 
Calibration here refers to the situation where you have added molecular spike-in DNA into your sample before library preparation and you want to leverage them to improve the quantification of your sample. <br/>

The scripts in this repository will allow you to:

1. Create a chimera genome from 2 fasta files (1 for the spike-in genome and 1 for the genome of interest)
2. Align your paired-end FastQ files to it 
3. Retain uniquely mapped (and deduplicated) reads 
4. Extract the corresponding DNA fragments from the paired reads, either uniquely mapping to the spike-in genome or the genome of interest
5. Obtain a scaled genome coverage for the fragments mapping to the genome of interest


**NB:** The scripts in this repository have been designed to be executed on the HPC. They have not yet been tested on a personal computer.

## Content

Below an overview of the content of this repository.

```
align-to-chimera/
├── config
│   └── modules.txt 
├── README.md
├── scripts
│   ├── calibrated-pe-alignment.sh
│   ├── create-chimera-genome.sh
│   └── parse-bowtie2-alignment-summary.R
└── slurm
```

* the `scripts/` folder contains all scripts necessary to execute the pipeline
* the `config/` folder contains a `.txt` file listing all modules (with versions) required to run the pipeline
* the `slurm/` folder contains two `.sh` files users can use to submit the pipeline either as an array or as a single sbatch command using SLURM on the HPC

## Quick start

Here's the step-by-step guide on how to run this pipeline:

* Copy this repository into your project folder 
* Ensure all scripts within the `align-to-chimera/scripts/` folder are executable and if not run `chmod +x /path/to/your/project/folder/align-to-chimera/scripts/*` from the command line
* Create a chimera genome assembly by running:

```
/path/to/your/project/folder/align-to-chimera/scripts/create-chimera-genome.sh -s <spike-in-genome> -g <genome-of-interest>  -o <outputDir> 
## Flags description
# -s = full path to the fasta file (including filename) of the spike-in genome. File can be gzip compressed and it must end with .fa.gz
# -g = full path to the fasta file (including filename) of the genome of interest. File can be gzip compressed and it must end with .fa.gz
# -o = the path to the directory where you want to save all the output of this script. Here the script will create 2 directories: 1) named <spike-in-genome>-<genome-of-interest> containing all output files and 2) a logs directory containing the log from bowtie2-build
```

* Align reads to the chimera genome by running:

```
/path/to/your/project/folder/align-to-chimera/scripts/calibrated-pe-alignment.sh  -o <outDir> -d <dataDir> -g <chimeraGenomeDir>
## Flags description
# -o = full path to the out directory where all final files and subdirectories will be saved
# -d = full path to the data directory containing your fastq files. Fastq files should be gz compressed and should end with <sample-name>_R1.fastq.gz and <sample-name>_R2.fastq.gz
# -s = sample name of the fastq files. This should correspond to the <sample-name> preceeding _R1.fastq.gz and _R2.fastq.gz. eg: if the fastq files are labelled as <your-sample>_R1.fastq.gz and <your-sample>_R2.fastq.gz, then -s your-sample.
# -g = full path to the directory containing all files (ie, bowtie2 index, chromosome.sizes.txt) for the chimera genome previously generated by the create-chimera-genome.sh script. NB: file naming for the chimera genome should be <spikeGenome-genomeOfInterest>, eg ecoliASM584v2-hg38
```
